--- linux-4.18/net/ipv4/tcp_ipv4.c	2018-08-12 20:41:04.000000000 +0000
+++ ../linux-4.18/net/ipv4/tcp_ipv4.c	2019-02-28 16:39:47.689440790 +0000
@@ -55,6 +55,7 @@
 #include <linux/bottom_half.h>
 #include <linux/types.h>
 #include <linux/fcntl.h>
+#include <linux/file.h>
 #include <linux/module.h>
 #include <linux/random.h>
 #include <linux/cache.h>
@@ -2005,6 +2006,11 @@ void tcp_v4_destroy_sock(struct sock *sk
 	tcp_fastopen_destroy_cipher(sk);
 	tcp_saved_syn_free(tp);

+	if (tp->tcp_monitor_ctx != NULL) {
+    eventfd_signal(tp->tcp_monitor_ctx, 1);
+    eventfd_ctx_put(tp->tcp_monitor_ctx);
+  }
+
 	sk_sockets_allocated_dec(sk);
 }
 EXPORT_SYMBOL(tcp_v4_destroy_sock);
